Add TCP Fast Open support

oerdnj opened the pull request:
This is a preliminary effort to add TCP Fast Open support to libuv. It compiles, but it might be completely broken at the moment.

I am submitting this as PR to get a feedback if this is heading into right direction, or it should be rewritten in a completely different way.

NawarA commented:
Is this still open? Should be reviewed and closed or merged

saghul commented on src/unix/stream.c:
```diff
@@ -837,6 +841,30 @@ static void uv__write(uv_stream_t* stream) {
 #endif
+#ifdef MSG_FASTOPEN
+  } else if (stream->flags & UV_TCP_FASTOPEN) {
+    struct sockaddr *addr = { 0 };
```
make this a `struct sockaddr_storage`, `struct sockaddr` is not large enough for an IPv6 address. Also allocate it on the stack.

saghul commented on src/unix/stream.c:
```diff
@@ -837,6 +841,30 @@ static void uv__write(uv_stream_t* stream) {
+      n = -1;
+    } else {
+      do {
+        n = sendto(uv__stream_fd(stream), iov[0].iov_base, iov[0].iov_len,
```
line these up vertically please

oerdnj replied:
What do you mean? I rechecked and it's aligned correctly here.

saghul commented on src/unix/stream.c:
```diff
@@ -837,6 +841,30 @@ static void uv__write(uv_stream_t* stream) {
+#  if defined(__APPLE__)
+      /*
+       * Due to a possible kernel bug at least in OS X 10.10 "Yosemite",
+       * EPROTOTYPE can be returned while trying to write to a socket that is
```
can you link to any bug reports / documentation about this?

oerdnj replied:
This is all over the code around, originally reported here: https://github.com/libuv/libuv/pull/482

saghul commented on src/unix/tcp.c:
```diff
@@ -163,27 +163,29 @@ int uv__tcp_connect(uv_connect_t* req,
-    /* If we get a ECONNREFUSED wait until the next tick to report the
-     * error. Solaris wants to report immediately--other unixes want to
-     * wait.
+  if (!(handle->flags & UV_TCP_FASTOPEN)) {
```
i think we should return an error instead. EISCONN perhaps?

oerdnj replied:
This would call for a special handling in the client.  Why would you want to do that?

saghul commented on src/unix/tcp.c:
```diff
@@ -319,6 +321,15 @@ int uv__tcp_nodelay(int fd, int on) {
+  if (setsockopt(fd, IPPROTO_TCP, TCP_FASTOPEN, &on, sizeof(on)))
+    return -errno;
+#endif
+  return 0;
```
this should return ENOTSUP or ENOSYS if it's not supported. The user needs to know.

saghul commented on src/unix/tcp.c:
```diff
@@ -359,6 +370,22 @@ int uv_tcp_nodelay(uv_tcp_t* handle, int on) {
+      return err;
+  }
+
+  if (qlen)
```
Let's not do this. If there is no socket, fail with EBADF. This is already troublesome with the keepalive options and I wouldn't want to make the same mistake.

saghul commented on src/win/tcp.c:
```diff
@@ -1246,6 +1246,10 @@ int uv_tcp_nodelay(uv_tcp_t* handle, int enable) {
 
 
+int uv_tcp_fastopen(uv_tcp_t* handle, int enable) {
+  return 0;
```
return UV_ENOTSUP.

saghul commented on docs/src/tcp.rst:
```diff
@@ -55,6 +55,11 @@ API
 
+.. c:function:: int uv_tcp_fastopen(uv_tcp_t* handle, int qlen)
+
+    Enable TCP Fast Open, it needs to be called befor calling uv_tcp_connect()
```
A link to at least the wikipedia article would be nice here.

txdv replied:
befor

sebdeckers commented:
LGTM

@saghul Any thoughts?

ewindisch commented on test/echo-server.c:
```diff
@@ -230,6 +230,15 @@ static int tcp4_echo_start(int port) {
+#ifdef TCP_FASTOPEN
+  r = uv_tcp_fastopen(&tcpServer, 5);
+  if (r) {
+    /* TODO: Error codes */
```
Are the TODOs intended to be fixed before merge?

oerdnj replied:
I guess, but I don't remember what I meant by this, it has been two years :)

oerdnj replied:
Ah, I see.  I am just mimicking the code around:
```
  r = uv_tcp_init(loop, &tcpServer);
  if (r) {
    /* TODO: Error codes */
    fprintf(stderr, "Socket creation error\n");
    return 1;
  }

  r = uv_tcp_bind(&tcpServer, (const struct sockaddr*) &addr, 0);
  if (r) {
    /* TODO: Error codes */
    fprintf(stderr, "Bind error\n");
    return 1;
  }
```

santigimeno commented on src/unix/stream.c:
```diff
@@ -845,21 +866,15 @@ static void uv__write(uv_stream_t* stream) {
-    if (errno != EAGAIN && errno != EWOULDBLOCK) {
+    if ((errno == EISCONN) && (stream->flags & UV_TCP_FASTOPEN)) {
+      /* The stream is already connected */
+      stream->flags &= ~UV_TCP_FASTOPEN;
```
In which scenario can this happen?

santigimeno commented on test/test-tcp-fastopen.c:
```diff
@@ -0,0 +1,247 @@
+/* Copyright Joyent, Inc. and other Node contributors. All rights reserved.
```
Change the header to something like [this](https://github.com/libuv/libuv/blob/v1.x/test/test-fs-copyfile.c#L1)?

santigimeno commented on test/test-tcp-fastopen.c:
```diff
@@ -0,0 +1,247 @@
+static int shutdown_cb_called = 0;
+static int connect_cb_called = 0;
+static int write_cb_called = 0;
+static int close_cb_called = 0;
```
No need to initialize static variables to zero

santigimeno commented on test/test-tcp-fastopen.c:
```diff
@@ -0,0 +1,247 @@
+    ASSERT(nread == 4);
+    ASSERT(memcmp("PING", buf->base, nread) == 0);
+  }
+  else {
```
style: `} else {`

santigimeno commented on Makefile.am:
```diff
@@ -242,6 +242,7 @@ test_run_tests_SOURCES = test/blackhole-server.c \
                          test/test-tcp-connect-error.c \
                          test/test-tcp-connect-timeout.c \
                          test/test-tcp-connect6-error.c \
+                         test/test-tcp-fastopen.c \
```
This file should be added to `uv.gyp` and `checksparse.sh` too

santigimeno commented on docs/src/tcp.rst:
```diff
@@ -55,6 +55,18 @@ API
+    initialized by `uv_tcp_init_ex` with address family set, so the
+    socket is created and before calling `uv_tcp_connect` for outgoing
+    connections or `uv_tcp_listen` for incoming connections.
+
```
We could add this link: https://lwn.net/Articles/508865/

santigimeno commented on include/uv-unix.h:
```diff
@@ -266,7 +266,8 @@ typedef struct {
   UV_STREAM_PRIVATE_PLATFORM_FIELDS                                           \
 
-#define UV_TCP_PRIVATE_FIELDS /* empty */
+#define UV_TCP_PRIVATE_FIELDS /* empty */				      \
```
it's not empty anymore

oyyd commented:
Hi, @oerdnj. Great work and it seems stalled now. Would you mind me picking this up and push forward basing on your work?

oerdnj commented:
@oyyd Sure, I would be delighted if somebody would finish the work into mergeable state. Thanks!

patrickkh7788 commented:
@oerdnj @oyyd  Any update ?  or maybe @oerdnj can write some guite to help people understand how to finihish this work.

oyyd commented:
@patrickkh7788 No. I found it was a lot more complicated than imagined :-(

oerdnj commented:
I might need it for another different DNS server I am working on, so I’ll take a look (or somebody from my team will).

--
Ondřej Surý <ondrej@sury.org>

> On 3 Mar 2019, at 13:00, Ouyang Yadong <notifications@github.com> wrote:
> 
> @patrickkh7788 No. I found it was a lot more complicated than imagined :-(
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

oerdnj commented on src/unix/stream.c:
```diff
@@ -826,16 +841,22 @@ static void uv__write(uv_stream_t* stream) {
+    }
+    do {
+      n = sendto(uv__stream_fd(stream), iov[0].iov_base, iov[0].iov_len,
+		 MSG_FASTOPEN, addr, addrlen);
```
I think I fixed that.

oerdnj commented on src/unix/stream.c:
```diff
@@ -407,6 +407,10 @@ int uv__stream_open(uv_stream_t* stream, int fd, int flags) {
       return -errno;
+
+    /* TODO Use qlen the user passed in. */
+    if ((stream->flags & UV_TCP_FASTOPEN) && uv__tcp_fastopen(fd, 5))
```
I still don't know what you meant by this.  Could you please elaborate a little bit on how do you think this should be better implemented?

oerdnj replied:
Oh, I already made the change :-):
```
    if ((stream->flags & UV_HANDLE_TCP_FASTOPEN)) {                             |
            return UV_EINVAL;                                                   |
    }
```

oerdnj commented:
JFTR The `MSG_FASTOPEN` is not available on macOS (although it defines `TCP_FASTOPEN`, and the test is broken because of that, so I have to check how to enable TFO on macOS yet.

rikuayanokozy commented:
@oerdnj macOS use `connectx` with flags `CONNECT_DATA_IDEMPOTENT` to indicate a socket use TFO.
See http://www.manpagez.com/man/2/connectx/ and
https://github.com/facebook/folly/blob/95ceac6552694ee1dc4f6f289a408aa9eeabc262/folly/detail/SocketFastOpen.cpp#L68

patrickkh7788 commented:
@oerdnj  Do you think I can use this patch on linux/android without problem ?

vcunat commented:
Well, the first iteration might be just a no-op on macOS, right?

sebdeckers commented:
Has TFO actually been enabled by default on any major browser (or OS)? FWIW, QUIC should have a similar feature.

Ref:
https://github.com/curl/curl/issues/3662#issuecomment-471315063

vcunat commented:
QUIC doesn't really seem related to this libuv change, except for the perspective of QUIC _perhaps_ lessening the desire to make TCP faster.

sebdeckers commented:
@vcunat I mention QUIC because the claimed difficulty with deploying TCP Fast Open is sometimes given as a reason behind the QUIC design. Similar benefits, different design. E.g. https://www.nanog.org/sites/default/files/Paasch_Network_Support.pdf

I'm not trying to argue against TFO. Merely providing context.

patrickkh7788 commented:
@oerdnj 

I try apply the patch to the recent  v1.30.1, seems not work anymore.  can you rebase this PR to made it mergeable into  v1.30.1?

patrickkh7788 commented:
@oerdnj 
I apply your patch ( https://gist.github.com/patrickkh7788/e8724c4424fb8c9fad45ca832e5e90ad )  on libuv v1.30.1,  when I call the uv_write, the data is send twice.  can you take a look at this ?

changloong commented:
@oerdnj 
Hi,   Ondřej Surý, thanks for the nice work.

I try cherry pick your branch into last release version, it work on my local server address.

when I try connect to a real nginx server.   the write callback return `UV_EALREADY` ( `connection already in progress`).  have you ever got this problem ?
 
The uv_tcp_connect callback has no error, and it cost 0ms to connected.  but when I try write I get `UV_EALREADY`.  (test on linux)

oerdnj commented:
I rebased the branch on top of current `v1.x` branch, cleaned up the code a little bit and pushed.  The commit message isn't up to par with current standard, but the code is there and it compiles.

It has been a long time since I first worked on this, so the only teeny-tiny problem is that the test fails, so it might not work at all :)

```
not ok 276 - tcp_fastopen_open
# exit code 6
# Output from process `tcp_fastopen_open`: (no output)
Output from process `tcp_fastopen_open`:
# uv_write error: address family not supported
# Assertion failed in test/test-tcp-fastopen.c on line 119: 0
```

oerdnj commented on src/unix/stream.c:
```diff
@@ -826,16 +841,22 @@ static void uv__write(uv_stream_t* stream) {
+		 MSG_FASTOPEN, addr, addrlen);
+    }
+    while (n == -1 && (ok_errno || errno == EINPROGRESS));
+    stream->flags &= ~UV_TCP_FASTOPEN;
```
The TFO applies only for the first message in the stream (as it sends the data along the handshake), for the subsequent packets, it doesn't make any difference and it's better to use the standard way of sending packets.

oerdnj replied:
The documentation on TCP FastOpen isn't the greatest, but I found this: https://dnsprivacy.org/wiki/display/DP/TCP+Fast+Open#TCPFastOpen-LinuxAPI

- On the client, the call to connect() is replaced with a call to sendmsg() or sendto() with the flags parameter set to MSG_FASTOPEN. 
  - For blocking sockets this performs both the handshake and sends the data before returning.

  - For non-blocking socket it returns the number of bytes buffered and sent in the SYN packet. 

    - If the cookie is available it will send the data before returning.

    - If the cookie is not available locally, it returns -1 with errno EINPROGRESS, and sends a SYN with TFO cookie request automatically. The caller needs to write the data again when the socket is connected with a call to send() or similar.

- Subsequent writes on this socket must also be done with send() or similar. A subsequent call with sendto() will return the error EISCONN.

oerdnj commented:
Ok, now the test if fixed.  The `sendto()` can return `EINPROGRESS` in which case it submitted a data-less SYN packet to get the cookie, so the `sendto()` needs to be repeated.  After the fix the test works as expected.

santigimeno commented on src/win/tcp.c:
```diff
@@ -1280,6 +1280,10 @@ int uv_tcp_nodelay(uv_tcp_t* handle, int enable) {
+int uv_tcp_fastopen(uv_tcp_t* handle, int enable) {
+  return UV_ENOTSUP;
+}
+
```
Apparently it's supported [starting with Windows 10, version 1607](https://docs.microsoft.com/en-us/windows/win32/winsock/ipproto-tcp-socket-options)

santigimeno commented on src/unix/tcp.c:
```diff
@@ -373,6 +380,27 @@ int uv__tcp_nodelay(int fd, int on) {
+ *
+ * Note on OS X the qlen MUST be 1 (the actual value is set via the
+ * net.inet.tcp.fastopen_backlog kernel parameter).
+ */
```
I would add this info about `os x` to the documentation

santigimeno commented on include/uv/unix.h:
```diff
@@ -295,7 +295,8 @@ typedef struct {
 
-#define UV_TCP_PRIVATE_FIELDS /* empty */
+#define UV_TCP_PRIVATE_FIELDS						      \
+  struct sockaddr_storage addr;
```
This field breaks ABI compatibility. IIRC, in previous versions you were calling `getsockname()` before calling `sendto()`, that might be a solution to avoid the breakage. Otherwise you should target the PR to the master branch.

tomaskrizek replied:
You probably mean `getpeername()` since the `addr` field is used to store the peer address in the `uv__tcp_connect()` call. However, I don't see a way to set the peer name without actually calling `connect()`.

Is there some other way to add the client-side TCP Fast Open support without breaking the ABI?

vtjnash replied:
Target at `master` branch is probably the main way (as then you are permitted to break the API), as there doesn't appear to be unused space elsewhere to fit this (assuming it does need to be stored).

vcunat replied:
:thinking: in that case, is there any rough idea when an ABI change might get into a release? (i.e. current master)  Even a couple years from now might not be too bad, by the looks of this PR's history.

santigimeno commented on Makefile.am:
```diff
@@ -265,6 +265,7 @@ test_run_tests_SOURCES = test/blackhole-server.c \
                          test/test-tcp-connect-error.c \
                          test/test-tcp-connect-timeout.c \
                          test/test-tcp-connect6-error.c \
+                         test/test-tcp-fastopen.c \
```
Add this file to `CMakelists.txt`

NawarA commented:
Almost there .. ;)

tomaskrizek commented on src/unix/stream.c:
```diff
@@ -886,6 +891,34 @@ static void uv__write(uv_stream_t* stream) {
+    do {
+      n = sendto(uv__stream_fd(stream), iov[0].iov_base, iov[0].iov_len,
+                 MSG_FASTOPEN, (const struct sockaddr*)addr, addrlen);
+    } while (n == -1 && (RETRY_ON_WRITE_ERROR(errno) || errno == EINPROGRESS));
```
The `EINPROGRESS` has to be handled differently than re-trying `sendto()` again (which only results in `EALREADY`).

I think when `EINPROGRESS` is returned, it should be handled similarly to when it happens in `uv__tcp_connect()`. The code as is doesn't seem to be able to handle subsequent writes to the stream until the connection is established.

sharadraju commented:
Any updates on this PR?
This will be very useful for us to add support for TCP Fast Open in Node.js

bnoordhuis commented:
My recollection (it's been a while) is that this PR is unlikely to get merged anytime soon because:

1. TFO is effectively Linux-only at the moment, and

2. Even on Linux it's not always usable/enabled/etc.

vcunat commented:
Libraries not supporting a feature because it's not commonly used?  Isn't that a circular argument?

oerdnj commented:
I’ve seen the discussion about TFO for njs, so I guess I can rebase this on top of current v1.x and try to get this merged again.

sharadraju commented:
> I’ve seen the discussion about TFO for njs, so I guess I can rebase this on top of current v1.x and try to get this merged again.

Great! Thanks @oerdnj . That would be really helpful!

bnoordhuis commented:
> Libraries not supporting a feature because it's not commonly used? Isn't that a circular argument?

No, because that's not the argument. Libuv is a platform abstraction layer. If Linux is the only platform that implements TFO, then it makes no sense to support that in libuv because there's nothing to abstract.

The fact that it doesn't even always work on Linux is just icing on the cake.

sreguna commented:
I see that Windows support is also available. https://learn.microsoft.com/en-us/windows/win32/api/Mswsock/nc-mswsock-lpfn_connectex

saghul commented:
TIL! This doesn't sound very promising, however:

```
Note that to make use of fast opens, you should use [ConnectEx](https://learn.microsoft.com/en-us/windows/desktop/api/Mswsock/nc-mswsock-lpfn_connectex) to make the initial connection, and specify data in that function's lpSendBuffer parameter to be transferred during the handshake process. Some of the data in lpSendBuffer will be transferred under the Fast Open protocol.
```

So seentially we could only support that if you call `uv_write` _before_ `uv_connect` actually does the connecting... sounds pretty clunky to use IMHO.

oerdnj commented:
> TIL! This doesn't sound very promising, however:
> 
> ```
> Note that to make use of fast opens, you should use [ConnectEx](https://learn.microsoft.com/en-us/windows/desktop/api/Mswsock/nc-mswsock-lpfn_connectex) to make the initial connection, and specify data in that function's lpSendBuffer parameter to be transferred during the handshake process. Some of the data in lpSendBuffer will be transferred under the Fast Open protocol.
> ```
> 
> So seentially we could only support that if you call `uv_write` _before_ `uv_connect` actually does the connecting... sounds pretty clunky to use IMHO.

There are already places with delayed syscalls, aren’t they? We can delay the connect system call until the first write happens.

saghul commented:
> There are already places with delayed syscalls, aren’t they? We can delay the connect system call until the first write happens.

That could work!

bnoordhuis commented:
> We can delay the connect system call until the first write happens.

Can you elaborate on how that would work? I can parse that sentence in multiple ways, none which seem feasible or desirable.

saghul commented:
I'd assume it means to do something like:

```
h = uv_tcp_init(...);
uv_tcp_fastopen(h, 1);
uv_connect(...)
uv_write(...)
# connect would happen when uv_write is called, on Windows at least
```

oerdnj commented:
> I'd assume it means to do something like:
> 
> ```
> h = uv_tcp_init(...);
> uv_tcp_fastopen(h, 1);
> uv_connect(...)
> uv_write(...)
> # connect would happen when uv_write is called, on Windows at least
> ```

Well, in fact, the PR already does that on Linux. It just have been so long, that I've forgotten what I did here :).

bnoordhuis commented:
Yeah, that was one of the possible interpretations. :-)

Okay, I see a couple of problems with that:

1. undesirable: also needs to special-case connect + read; more special cases == more ugly

2. undesirable: breaks code that only starts writing from uv_connect_cb; makes uv_tcp_fastopen() not a drop-in replacement

3. undesirable: delaying the connect() call regresses handshake performance

IMO, it'd be best to go in the write-before-connect direction. Clunky but no other undesirable attributes.

oerdnj commented:
> Yeah, that was one of the possible interpretations. :-)
> 
> Okay, I see a couple of problems with that:
> 
> 1. undesirable: also needs to special-case connect + read; more special cases == more ugly

The only special case needed would be that you need to start writing (`uv_write()`) before you start reading (`uv_read_start()`), otherwise TFO would be ineffective.

Note: `uv_try_write()` should always fail in not connected yet, so we can process the connect failures.

> 2. undesirable: breaks code that only starts writing from uv_connect_cb; makes uv_tcp_fastopen() not a drop-in replacement

Why? The `uv_connect()` would call the callback immediately (without calling system `connect()`), so in fact, it would be a drop in replacement.

> 3. undesirable: delaying the connect() call regresses handshake performance

Can you elaborate a bit more? I don't understand what you mean here.

> IMO, it'd be best to go in the write-before-connect direction. Clunky but no other undesirable attributes.

bnoordhuis commented:
Pretty much all downstream code currently is structured like this:
```c++
uv_connect(connect_req, handle, [](...) {
  uv_write(...);
  // or:
  uv_read_start(...);
});
```
That won't work if  the connect req doesn't do anything until the first uv_read_start() or uv_write().

Apropos:

> undesirable: delaying the connect() call regresses handshake performance

Example: Node's http connection pool opens a bunch of connections to the host but doesn't do anything with them until there's a request ready to be sent.

At the moment `uv_tcp_connect()` calls connect() straight away. The kernel performs the TCP handshake in the background. The connections are ready by the time they're needed.

With your proposal, the connect() call isn't done until the request is sent. That adds latency.
